### ISSUE TITLE
My sonar does not recognize thes pattern ** in includeDirectorie propertie

### ISSUE THREAD

**Issue by icaro56** [SonarOpenCommunity/sonar-cxx#1601]
My sonar does not recognize thes pattern below. But in the propertie exclusion works (e.g. sonar.exclusions=\**/GeneratedFiles/**/*, \**/GeneratedFiles/**, \**/*GeneratedFiles/**/*, Temp/**)

sonar.cxx.includeDirectories=Private, Public, Public/**/*.h

In Public/**/*.h the error occurs:

Start\Public***.h'
java.io.IOException: Invalid argument
at java.io.WinNTFileSystem.canonicalize0(Native Method)
at java.io.WinNTFileSystem.canonicalize(WinNTFileSystem.java:428)
at java.io.File.getCanonicalPath(File.java:618)
at java.io.File.getCanonicalFile(File.java:643)
at org.sonar.cxx.preprocessor.SourceCodeProvider.setIncludeRoots(SourceCodeProvider.java:53)
at org.sonar.cxx.preprocessor.CxxPreprocessor.(CxxPreprocessor.java:249)
at org.sonar.cxx.preprocessor.CxxPreprocessor.(CxxPreprocessor.java:235)
at org.sonar.cxx.parser.CxxParser.create(CxxParser.java:66)
at org.sonar.cxx.CxxAstScanner.create(CxxAstScanner.java:101)

* SonarQube version? SonarQube Scanner 3.0.3.778

11:06:10.521 DEBUG: Plugins:
11:06:10.521 DEBUG:   * C++ (Community) 0.9.8 (cxx)
11:06:10.521 DEBUG:   * SonarPython 1.8.0.1496 (python)
11:06:10.521 DEBUG:   * SonarQube :: Plugins :: SCM :: SVN 1.6.0.860 (scmsvn)
11:06:10.521 DEBUG:   * SonarJS 3.3.0.5702 (javascript)
11:06:10.521 DEBUG:   * SonarC# 6.6.0.3969 (csharp)
11:06:10.521 DEBUG:   * SonarJava 4.15.0.12310 (java)
11:06:10.521 DEBUG:   * Flex 2.3 (flex)
11:06:10.521 DEBUG:   * SonarQube :: Plugins :: SCM :: Git 1.3.0.869 (scmgit)
11:06:10.521 DEBUG:   * SonarXML 1.4.3.1027 (xml)
11:06:10.521 DEBUG:   * SonarPHP 2.11.0.2485 (php)
11:06:10.521 DEBUG:   * SonarTS 1.1.0.1079 (typescript)
11:06:10.521 DEBUG:   * Coverity 1.6.2 (coverity)

**Comment by guwirth** [comment]
@icaro56 please try without blanks after the commas:

**sonar.cxx.includeDirectories=Private,` `Public,` `Public/**/*.h**

**Comment by icaro56** [comment]
I tested without blank spaces but the error did not fix

**Comment by guwirth** [comment]
@icaro56 oh sorry, `sonar.cxx.includeDirectories` does not support patterns. For includes there must be a fixed predefined order.

https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Supported-configuration-properties#sonarcxxincludedirectories

**Comment by icaro56** [comment]
If the pattern was accepted, it would make my life much easier. :)

**Comment by guwirth** [comment]
@icaro56 what exactly is your use case?

`sonar.cxx.includeDirectories=Private, Public, Public/**/*.h`

* First of all this is a list of directories. So *.h can't work.
* You can add relative (to SQ project root)  or absolute paths.
* These paths are used to preprocess the files. So same as with other compilers you have to set them all and in the right order.
* What should a placeholder do? In which order should he search for includes? Seems to be more or less random?

**Comment by icaro56** [comment]
I would like the include all folder recursivelly. Maybe thus: Public/**

Or, I wish I could ignore these warnings regarding the includes. Because when I add the QT includes, visual studio and windows sdk, the execution time is taking about 30 minutes

**Comment by guwirth** [comment]
@icaro56 

> Or, I wish I could ignore these warnings regarding the includes. 

Please try a newer version (1.2) of the plugin. You get less warnings there.
Missing includes can be ignored as long you get not too much parsing errors.
Typically you can fix such parsing errors by add adding a forceinclude file with some macros to support the parser. https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Dealing-with-compiler-specific-code-pieces
This should also speed-up the analysis time.

> Maybe thus: Public/**

Problem is that the order the preprocessor is looking for includes is then random. It would take the first one in the filesystem (sorted by ...?).

**Comment by dl5rcw** [comment]
I agree to @icaro56 explanation. Using Qt but any other external Library, it will take execution time up. Typically, we want to ignore include issues of any SDK used in the project, not add them to the project.

**Comment by guwirth** [comment]
@icaro56 & @dl5rcw we are always open for good proposals, that's the reason I opened the issue again.

* The problem with the (many) warnings: should be solved with v1.2. We reduced the number of warnings.
* Why does the C++ plugin parse the code? This is needed to create code metrics and for some internal checks based on the from the parser created AST. In case you use external code checkers and are not interested in code metrics you should not set any include directories and ignore all parsing warnings.
* The problem with the includes:
  * Adding includes means at the end always more to parse and slower analysis times. Adding patterns to the include directories makes this even worse because the preprocessor has to check much more locations. General recommendation is **not to add** includes of external libraries (Boost, Qt, Xerces, ATL, MFC, ...). 
   * On the other hand some includes or at least macros are needed to modify the code to generate standard C++ grammar. It's important to understand, that the plugin checks only against the grammar. Means as long the result is valid C++ code everything works.
   * The easiest solution is here to mock away the library specific extensions:  https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Dealing-with-compiler-specific-code-pieces

One simple sample:
```C++
EXTERN void CALLCONV sample(UINT32 p) {
}
```
Can't be parsed. In case you help a little bit:
```C++
#define EXTERN
#define CALLCONV
EXTERN void CALLCONV sample(UINT32 p) {
}
```
this results after preprocessing in:
```C++
void sample(UINT32 p) {
}
```
Which works fine. As you can see you don't have to define data types. The grammar is able to handle them also if they are unknown. A good tool to get a feeling is the SSLR toolkit. Here you can add samples and play around: https://github.com/SonarOpenCommunity/sonar-cxx/releases/download/cxx-1.2.0/sslr-cxx-toolkit-1.2.0.jar

With Qt you have to find most likely a solutions for the specific [Qt macros](http://doc.qt.io/qt-5/qtglobal.html). A sample for Visual Studio is here: https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/resources/macros/VS10Macros.h.

Looking to Cppcheck in [qt.cfg](https://github.com/danmar/cppcheck/blob/master/cfg/qt.cfg) they are mocking:
```
  <define name="Q_DECL_EXPORT" value=""/>
  <define name="Q_DECL_IMPORT" value=""/>
  <define name="Q_DECLARE_FLAGS(x,y)" value=""/>
  <define name="Q_DISABLE_COPY(C)" value="C(C&amp;);C&amp; operator=(const C&amp;);"/>
  <define name="Q_ENUM(X)" value=""/>
  <define name="Q_ENUMS(X)" value=""/>
  <define name="Q_FLAGS(X)" value=""/>
  <define name="Q_INTERFACES(X)" value=""/>
  <define name="Q_OBJECT" value=""/>
  <define name="Q_PROPERTY(X)" value=""/>
  <define name="foreach(A,B)" value="for(A:B)"/>
  <podtype name="qint8" sign="s" size="1"/>
  <podtype name="qint16" sign="s" size="2"/>
  <podtype name="qint32" sign="s" size="4"/>
  <podtype name="qint64" sign="s" size="8"/>
  <podtype name="quint8" sign="u" size="1"/>
  <podtype name="quint16" sign="u" size="2"/>
  <podtype name="quint32" sign="u" size="4"/>
  <podtype name="quint64" sign="u" size="8"/>
  <podtype name="qlonglong" sign="s"/>
  <podtype name="qulonglong" sign="u"/>
  <podtype name="uchar" sign="u"/>
  <podtype name="uint" sign="u"/>
  <podtype name="ulong" sign="u"/>
  <podtype name="ushort" sign="u"/>
  <podtype name="qptrdiff" sign="s"/>
  <podtype name="quinitptr" sign="u"/>
```
Think that's a good starting point.
